tests/test_readout.py

```diff
@@ -229,4 +244,151 @@ def test_compare_readout_methods(
+            10000,
+            chain(kron(RX(0, np.pi / 6), RX(1, np.pi / 4)), CNOT(0, 1)),
+            BackendName.PYQTORCH,
+        ),
+    ],
+)
+def test_readout_mthree_mitigation(
+    error_probability: float,
+    n_shots: int,
+    block: AbstractBlock,
+    backend: BackendName,
+) -> None:
+    circuit = QuantumCircuit(block.n_qubits, block)
+    noise = Noise(protocol=Noise.READOUT, options={"error_probability": error_probability})
+
+    model = QuantumModel(circuit=circuit, backend=backend)
+
+    ordered_bitstrings = [bin(k)[2:].zfill(block.n_qubits) for k in range(2**block.n_qubits)]
+
+    mitigation_mle = Mitigations(
+        protocol=Mitigations.READOUT,
+        options={"optimization_type": ReadOutOptimization.MLE, "n_shots": n_shots},
+    ).mitigation()
+
+    samples_mle = mitigation_mle(model=model, noise=noise)[0]
+    p_mle = np.array([samples_mle[bs] for bs in ordered_bitstrings]) / sum(samples_mle.values())
+
+    mitigation_mthree = Mitigations(
+        protocol=Mitigations.READOUT,
+        options={"optimization_type": ReadOutOptimization.MTHREE, "n_shots": n_shots},
+    ).mitigation()
+    samples_mthree = mitigation_mthree(model=model, noise=noise)[0]
+    p_mthree = np.array([samples_mthree[bs] for bs in ordered_bitstrings]) / sum(
+        samples_mthree.values()
+    )
+
+    assert wasserstein_distance(p_mle, p_mthree) < LOW_ACCEPTANCE
+
+
+def test_readout_mthree_sparse() -> None:
+    n_qubits = 10
+    exact_prob = np.random.rand(2 ** (n_qubits - 5))
+    exact_prob = exact_prob / sum(exact_prob)
+    exact_prob = np.concatenate([exact_prob, np.zeros(2**n_qubits - len(exact_prob))], axis=0)
+    exact_prob = 0.90 * exact_prob + 0.1 * np.ones(2**n_qubits) / 2**n_qubits
+    np.random.shuffle(exact_prob)
+
+    observed_prob = np.array(exact_prob, copy=True)
+    observed_prob[exact_prob < 1 / 2 ** (n_qubits)] = 0
+
+    noise_matrices = []
+    for t in range(n_qubits):
+        t_a, t_b = np.random.rand(2) / 8
+        K = np.array([[1 - t_a, t_a], [t_b, 1 - t_b]]).transpose()  # column sum be 1
+        noise_matrices.append(K)
+
+    confusion_matrix_subspace = normalized_subspace_kron(noise_matrices, observed_prob.nonzero()[0])
+
+    input_csr = csr_matrix(observed_prob, shape=(1, 2**n_qubits)).T
+
+    p_corr_mthree_gmres = gmres(confusion_matrix_subspace, input_csr.toarray())[0]
+    p_corr_mthree_gmres_mle = mle_solve(p_corr_mthree_gmres)
+
+    noise_matrices_inv = list(map(matrix_inv, noise_matrices))
+    p_corr_inv_mle = mle_solve(tensor_rank_mult(noise_matrices_inv, exact_prob))
+
+    assert wasserstein_distance(p_corr_mthree_gmres_mle, p_corr_inv_mle) < LOW_ACCEPTANCE
+
+
+@pytest.mark.flaky(max_runs=2)
+@pytest.mark.parametrize(
+    "n_qubits,index",
+    [
+        (4, 6),
+        (4, 1),
+        (5, 25),
+        (5, 20),
+        (5, 9),
+        (5, 11),
+        (4, 15),
+        (5, 7),
+    ],
+)
+def test_readout_majority_vote(n_qubits: int, index: int) -> None:
```

Comment:
Last point, if you could implement using the protocol as we discussed, that'd be great. Thanks.